tools/offline_log_viewer/kafka.py

```diff
@@ -32,18 +42,98 @@ def decode(self):
-                        "control_batch"]
-                    record_dict = record.kv_dict()
-                    records.append(record_dict)
-                    if is_tx_ctrl:
-                        record_dict["type"] = self.get_control_record_type(
-                            record.key)
-                    yield record_dict
+            if self.headers_only:
+                continue
+            records = batch
+            if header["type_name"] == "tm_update":
+                records = list(batch)
+                assert len(records) == 1
+            for record in records:
+                attrs = header["expanded_attrs"]
+                is_tx_ctrl = attrs["transactional"] and attrs["control_batch"]
+                record_dict = record.kv_dict()
+                if is_tx_ctrl:
+                    record_dict["type"] = self.get_control_record_type(
+                        record.key)
+                if header["type_name"] == "tm_update":
+                    size = struct.unpack("<i", record.key[9:13])[0]
+                    record_dict["key"] = {
+                        "type": int.from_bytes(record.key[0:1], "little"),
+                        "pid.id": struct.unpack("<q", record.key[1:9])[0],
+                        "tx.id": record.key[13:13 + size].decode("UTF-8")
+                    }
+                    record_dict["value"] = {
+                        "version": int.from_bytes(record.value[0:1], "little")
+                    }
+                    size = struct.unpack("<i", record.value[1:5])[0]
+                    record_dict["value"]["tx.id"] = record.value[5:5 +
+                                                                 size].decode(
+                                                                     "UTF-8")
+                    record_dict["value"]["pid"] = {
+                        "id":
+                        struct.unpack("<q",
+                                      record.value[5 + size:13 + size])[0],
+                        "epoch":
+                        struct.unpack("<h",
+                                      record.value[13 + size:15 + size])[0],
+                    }
+                    record_dict["value"]["tx_seq"] = struct.unpack(
+                        "<q", record.value[15 + size:23 + size])[0]
+                    record_dict["value"]["etag"] = struct.unpack(
+                        "<q", record.value[23 + size:31 + size])[0]
+                    record_dict["value"]["status"] = TxStatus(
+                        struct.unpack("<i", record.value[31 + size:35 +
+                                                         size])[0]).name
+                    record_dict["value"]["timeout_ms"] = struct.unpack(
+                        "<q", record.value[35 + size:43 + size])[0]
+                    record_dict["value"]["last_update_ts"] = struct.unpack(
+                        "<q", record.value[43 + size:51 + size])[0]
```

Comment:
Can we clean this up so we avoid the explicit buffer slicing math? The key to cleaning this up is to use the `struct` module to the slicing for us. Any fixed size concatenated values can be unpacked as a vector.

```
fields = struct.unpack("<qqiqq")
tx_seq = fields[0]
etag = fields[1]
...
```

but I would think that you should be able to use our `reader.py` for a lot of this, or a byte stream to avoid having to also track all the byte pointers (e.g. `o` below).

Reply:
good idea the code looks nicer now